### PR TITLE
Adding sample files to demonstrate how cloud providers leverage CCM

### PR DESCRIPTION
**Pull request by cici37**
**What type of PR is this?**
/kind cleanup

**What this PR does / why we need it**:
Added an basic working sample, an advanced sample and an detailed sample to demonstrate how cloud providers leverage CCM

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

File: staging/src/k8s.io/cloud-provider/sample/basic_main.go [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+
+	"github.com/spf13/pflag"
+	"k8s.io/cloud-provider"
+	"k8s.io/cloud-provider/app"
+	"k8s.io/cloud-provider/options"
+	"k8s.io/component-base/cli/flag"
+	"k8s.io/component-base/logs"
+	_ "k8s.io/component-base/metrics/prometheus/clientgo" // load all the prometheus client-go plugins
+	_ "k8s.io/component-base/metrics/prometheus/version"  // for version metric registration
+	"k8s.io/klog/v2"
```
Comment (cheftako) [inline staging/src/k8s.io/cloud-provider/sample/basic_main.go:38]: I think we need a comment here to indicate the cloud provider should import their cloud provider implementation?

File: staging/src/k8s.io/cloud-provider/sample/basic_main.go [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+)
+
+const (
+	// sampleCloudProviderName shows an sample of using hard coded parameter for CloudProviderName
```
Comment (cheftako) [inline staging/src/k8s.io/cloud-provider/sample/basic_main.go:42]: Add instructions in the comment for the cloud provider to replace these values?

File: staging/src/k8s.io/cloud-provider/sample/basic_main.go (hunk 1 above)
Comment (cheftako) [reply staging/src/k8s.io/cloud-provider/sample/basic_main.go:38]: Also please add these sort of comments in the advanced sample.

Comment (cheftako) [reply staging/src/k8s.io/cloud-provider/sample/basic_main.go:38]: Can we add a comment/link in the version in K/K private pointing to these?

Comment (cheftako) [reply staging/src/k8s.io/cloud-provider/sample/basic_main.go:38]: Also a simple README.md file in this directory explaining for "new" cloud providers that they should copy the sample into their repo and then import the rest of the code. Also pointing to the examples of how our extensibility works.

Comment (cici37) [reply staging/src/k8s.io/cloud-provider/sample/basic_main.go:38]: Added more comments. Will add the README.md file later as well. Thanks

File: staging/src/k8s.io/cloud-provider/sample/basic_main.go (hunk 2 above)
Comment (cici37) [reply staging/src/k8s.io/cloud-provider/sample/basic_main.go:45]: Added more comments

**Comment by cici37** [comment]
Possible issue related? https://github.com/kubernetes/test-infra/issues/19910

**Comment by cheftako** [comment]
> Possible issue related? [kubernetes/test-infra#19910](https://github.com/kubernetes/test-infra/issues/19910)

Quite possibly, @spiffxp might be able to confirm.
Makes me glad you merged the primary fix and left this as a follow up.

File: cmd/cloud-controller-manager/main.go [hunk 3]
```diff
@@ -41,16 +44,13 @@ import (
 	genericcontrollermanager "k8s.io/controller-manager/app"
 	"k8s.io/klog/v2"
 	nodeipamconfig "k8s.io/kubernetes/pkg/controller/nodeipam/config"
+	// For existing cloud providers, the option to import legacy providers is still available.
```
Comment (cheftako) [inline cmd/cloud-controller-manager/main.go:47]: So its not just in-tree providers. Out of tree providers can do the same things, however the import would obviously be to where ever they have their cloud provider implementation. For example Huawei has their implementation in https://github.com/kubernetes-sigs/cloud-provider-huaweicloud/tree/master/pkg/cloudprovider/huaweicloud.

File: staging/src/k8s.io/cloud-provider/sample/README.md [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+## Steps cloud providers shoud follow
+
+1. Have your external repo under k8s.io. e.g. `k8s.io/cloud-provider-<provider>`
+2. Create `main.go` file under your external repo CCM directory. Please refer to `basic_main.go` for a minial working sample and `advanced_main.go` for advanced configuration samples.
+Note: If you have a requirement of adding/deleting controllers within CCM, please refer to `k8s.io/kubernetes/cmd/cloud-controller-manager/main.go` for detailed samples.
+3. Build/release CCM from your external repo. For existing cloud providers, the option to import legacy providers from `k8s.io/legacy-cloud-provider/<provider>` is still available.
+
+## Things you should NOT do
+
+ 1. Vendor in `k8s.io/cmd/cloud-controller-manager`.
```
Comment (cheftako) [inline staging/src/k8s.io/cloud-provider/sample/README.md:18]: They should not need to vendor in any k8s.io/kubernetes/*.

Comment (cici37) [reply staging/src/k8s.io/cloud-provider/sample/README.md:18]: This is under `Things you should NOT do` ^^

**Pull request merged**